File: 91-decode-ways/decode-ways.go
var memo map[int]int

func getNumWays(s *string, index int) int {
    // memoize:
    if val, ok := memo[index]; ok{
        return val
    }

    // Base Cases: found a way
    if index == len(*s) {
        return 1
    }
    // leading char: 0 or >= 3
    firstChar:= (*s)[index] - '0'
    if firstChar == 0{
        return 0
    }

    numWays := 0
    // Choose only firstChar
    numWays += getNumWays(s, index + 1)
    // Choose both chars
    if index + 1 < len(*s){
        nextChar := (*s)[index + 1] - '0'
        if (*s)[index] == '1' || ((*s)[index] == '2' && nextChar <= 6) {
            numWays += getNumWays(s, index + 2)
        }
    }
    
    memo[index] = numWays
    return numWays

}


func numDecodings(s string) int {
    // leading zeros, cannot decode
    if len(s) == 0 || s[0] == '0'{
        return 0
    }
    memo = make(map[int]int, len(s) + 1)
    return getNumWays(&s, 0)

}

File: 91-decode-ways/decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        """
        6 0 6
        

        """
        # memo = {}
        # def count_ways(index: int) -> int:
        #     if index in memo:
        #         return memo[index]
            
        #     # Base Case
        #     if index == len(s):
        #         # reached end of string 1 way to decode
        #         return 1 
        #     if int(s[index]) == 0:
        #         # no way to decode 
        #         return 0

        #     num_ways = 0
        #     # 1 digit way 
        #     num_ways += count_ways(index + 1)
        #     if index + 1 < len(s) and (int(s[index]) * 10 + int(s[index + 1])) <= 26:
        #         num_ways += count_ways(index + 2)
        #     memo[index] = num_ways
        #     return num_ways
        
        dp = [0] * (len(s) + 1)
        dp[-1] = 1

        for i in range(len(s) - 1, -1, -1):
            if int(s[i]) == 0:
                dp[i] = 0
                continue
            dp[i] += dp[i + 1]
            if i + 1 < len(s) and (int(s[i]) * 10 + int(s[i + 1])) <= 26:
                dp[i] += dp[i + 2]
        
        return dp[0]


            